src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java

```diff
@@ -229,18 +231,40 @@ private List<ClusteringElements> bindAndGetClusteringElements(QueryOptions optio
         {
             case SINGLE_COLUMN:
             case TOKEN:
-                return bindAndGet(options).stream()
-                                          .map(b ->  ClusteringElements.of(columnsExpression.columnSpecification(), b))
-                                          .collect(Collectors.toList());
+                return bindAndGetSingleTermClusteringElements(options);
             case MULTI_COLUMN:
-                return bindAndGetElements(options).stream()
-                                                  .map(buffers -> ClusteringElements.of(columnsExpression.columns(), buffers))
-                                                  .collect(Collectors.toList());
+                return bindAndGetMultiTermClusteringElements(options);
             default:
                 throw new UnsupportedOperationException();
         }
     }
 
+    private List<ClusteringElements> bindAndGetSingleTermClusteringElements(QueryOptions options)
+    {
+        List<ByteBuffer> values = bindAndGet(options);
+        if (!values.isEmpty())
+        {
+            List<ClusteringElements> elements = new ArrayList<>(values.size());
+            for (ByteBuffer value : values)
```

Comment:
If we are concerned about allocations at this place, we should use a `for (int i = , m = values.size(); i < m; i++)` loop. We replaced those loops in the past already for the same reason.